theme-kit/router.ts

```diff
@@ -0,0 +1,147 @@
+/* eslint-disable @typescript-eslint/no-use-before-define,@typescript-eslint/naming-convention */
+
+import type { Culture } from '@prezly/sdk';
+import { notFound } from 'next/navigation';
+import type { ReactElement, ReactNode } from 'react';
+import UrlPattern from 'url-pattern';
+
+import { assertServerEnv } from '@/utils';
+
+import { api } from './api';
+
+export function route<
+    Pattern extends `/${string}`,
+    Match extends ExtractPathParams<Pattern>,
+    Props = Match,
+>(pattern: Pattern, loader: Loader<Match, Props>) {
+    const urlPattern = buildUrlPattern(pattern);
+
+    return {
+        async match(path: string, getDefaultLocale, isEnabledLocale) {
+            const matched = urlPattern.match(path);
+
+            if (!matched) {
+                return undefined;
+            }
+
+            if (matched.locale && !(await isEnabledLocale(matched.locale))) {
+                return undefined;
+            }
+
+            if (!matched.locale) {
+                matched.locale = await getDefaultLocale();
+            }
+
+            const { match: pageMatch, resolveLocale, Layout, default: Page } = await loader();
+
+            const pageMatched = pageMatch ? await pageMatch(matched) : matched;
+
+            if (!pageMatched) {
+                return undefined;
+            }
+
+            if (resolveLocale) {
+                return {
+                    match: {
+                        ...pageMatched,
+                        locale: await resolveLocale(pageMatched),
+                    },
+                    Layout,
+                    Page,
+                };
+            }
+
+            return {
+                match: { ...pageMatched, locale: matched.locale ?? (await getDefaultLocale()) },
+                Layout,
+                Page,
+            };
+        },
+    } as Route<Match, Props | Match>;
+}
+
+type Route<Match, Props> = {
+    match(
+        path: string,
+        getDefaultLocale: () => Promise<string>,
+        isEnabledLocale: (locale: string) => Promise<boolean>,
+    ): Promise<
+        | {
+              match: Match;
+              Layout?: ServerComponentType<Props>;
+              Page: ServerComponentType<Props>;
+          }
+        | undefined
+    >;
+};
+
+export async function match(path: string | string[], routes: Route<any, any>[]) {
+    assertServerEnv('match');
+
+    const { contentDelivery } = api();
```

Comment:
I found this shorthand naming a bit weird. Maybe I'm too used to `getX` naming for such methods 🤔

Reply:
> Maybe I'm too used to `getX` naming for such methods 🤔

This is a matter of preference. The academic Java-way of naming everything as `getX` is one approach. Next.js, for example, is preferring shorter names: `headers()` instead of `getRequestHeaders()`, same for `cookies()`. 

I wanted it to be compact and feel native to a Next.js application. And after you adjust the expectations, it looks good, if you ask me. :)